### ISSUE TITLE
Explicit padding for Transpose Convolution fails

### ISSUE THREAD

**Issue by riestmo-nxp** [ARM-software/armnn#751]
Hi,

I was trying to run a network which contains a TransposeConv layer with an explicit output shape as input and VALID padding (see screenshot). 

<img width="812" alt="image" src="https://github.com/ARM-software/armnn/assets/149694071/a77d13bb-2d41-4f4e-848d-19d3621c8b0b">

Unfortunately, I run into the following assertion:

`Assertion (expectedOutputShape == outputShapeGivenAsInput)`

I tried to trace back the issue a bit and spotted that the `TfLiteParser` skips the explicit padding when VALID padding is given, even if the output_shape is explicitly given:

```cpp
// Function that calculates explicit padding when the output shape is known.
// At the moment the output is only given as an input parameter in Transpose Convolution,
// not in Convolution and Depthwise Convolution
void CalcPadding(uint32_t inputSize,
                 uint32_t filterSize,
                 uint32_t stride,
                 uint32_t dilation,
                 uint32_t& paddingFront,
                 uint32_t& paddingBack,
                 tflite::Padding padding,
                 uint32_t outputSize)
{
    IgnoreUnused(dilation);
    paddingFront = 0;
    paddingBack = 0;
    if (padding == tflite::Padding_SAME)
    {
        uint32_t totalPadding = (inputSize - 1) * stride + filterSize - outputSize;
        paddingFront = totalPadding / 2;
        paddingBack = totalPadding - paddingFront;
    }
}
```

From my point of view the explicit padding should always be calculated and not only when padding SAME is specified for the layer. Or is my understanding wrong here?

Thanks for helping me out here!

**Comment by TeresaARM** [comment]
Hi @riestmo-nxp ,
if you want to attach the model that would help us to debug the issue.

**Comment by riestmo-nxp** [comment]
Hi @TeresaARM,
sure, that's the model: [model.zip](https://github.com/ARM-software/armnn/files/13816486/model.zip)

**Comment by xiaotongnii** [comment]
It seems that inferShape erro happened when Convert this model to armnn ir. And TfliteParser maybe not consider that VALD Padding mode.
This mode datalayout is NHWC or NCHW?
about TransposeConv padding mode :https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_transpose

**Comment by riestmo-nxp** [comment]
Hi @xiaotongnii,
the data format is NHWC (as usual in TFLite).

**Comment by TeresaARM** [comment]
Hello,
TF accepts both outputs shapes as valid: [1, 66, 226, 16] and [1, 65, 226, 16]

```
input  = tf.constant(1, shape=[1, 32, 224, 16]) # NHWCin
kernel = tf.constant(2, shape=[3, 3, 16, 16]) # HWCoutCin

x = tf.nn.conv2d_transpose(input,
                           kernel,
                           output_shape=[1, 66, 226, 16], #NHWCout
                           strides=(2, 1),
                           padding="VALID",
                           data_format='NHWC',
                           dilations=None,
                           name=None)

x.shape
```

```
input  = tf.constant(1, shape=[1, 32, 224, 16]) # NHWCin
kernel = tf.constant(2, shape=[3, 3, 16, 16]) # HWCoutCin

x = tf.nn.conv2d_transpose(input,
                           kernel,
                           output_shape=[1, 65, 226, 16], #NHWCout
                           strides=(2, 1),
                           padding="VALID",
                           data_format='NHWC',
                           dilations=None,
                           name=None)

x.shape
```
In ArmNN we calculate the output this way:
Out =  ( In - 1 ) * stride + filter - padding.
65   = ( 32 - 1 ) * 2        + 3      -  0

I am trying to see how we can support both output shapes, as TF does.

**Comment by xiaotongnii** [comment]
@riestmo-nxp I try convert model.tflite to model.armnn and everythins is OK! 
` ./ArmnnConverter -f tflite-binary -m ./model.tflite -i serving_default_input:0 -s 1,512,256,32 -o PartitionedCall:0 -p ./model.armnn`
armnn22.02 test OK，check your armnn version pls！

**Comment by xiaotongnii** [comment]
![Uploading transposeConv.png…]()

**Comment by xiaotongnii** [comment]
<img width="960" alt="IMG_3278" src="https://github.com/ARM-software/armnn/assets/44741734/02667f57-281c-4d0e-ae79-75707fd327e1">

**Comment by tracyn-arm** [comment]
Tflite padding is more complex than what we have in Arm NN. Take a look at:
https://github.com/tensorflow/tensorflow/blob/be8160ab0a925311041f7a53b7c7c148a3e51f35/tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc#L1206
In the `VALID` case, the calculations are:
```
        *padding_top = *padding_bottom = *padding_left = *padding_right = 0;
        *adjustment_height = (output_height - kernel_height) % stride_height;
        *adjustment_width = (output_width - kernel_width) % stride_width;
```
The `CalcPadding() ` code inserted above reflects the first line.  The difference occurs with the adjustment calculations. From the tflite code:
```
/// @param adjustment_height - additional elements in the bottom of the 2D output data.
/// @param adjustment_width - additional elements to the right of the 2D output data.
```
In the case that is failing with the provided model, the `adjustment_height = 1 # (64 - 3) % 2 , adjustment_width = 0 `

We use the `TransposeConv2d` from the Compute Library and I could not find support for the adjustment parameters unfortunately and this limits the type of padding we can have in Arm NN.

It's a complex problem so please let me know if you think I might have missed something.  Thanks.
